engine/runtime/src/main/scala/org/enso/compiler/Cache.scala

```diff
@@ -0,0 +1,463 @@
+package org.enso.compiler
+
+import com.oracle.truffle.api.{TruffleFile, TruffleLogger}
+import org.bouncycastle.jcajce.provider.digest.SHA3
+import org.bouncycastle.util.encoders.Hex
+import org.enso.interpreter.runtime.EnsoContext
+import org.enso.logger.masking.MaskedPath
+
+import java.io.{
+  ByteArrayInputStream,
+  ByteArrayOutputStream,
+  IOException,
+  ObjectInputStream,
+  ObjectOutputStream
+}
+import java.nio.file.{
+  FileAlreadyExistsException,
+  NoSuchFileException,
+  Path,
+  StandardOpenOption
+}
+import java.security.MessageDigest
+import java.util.logging.Level
+import scala.util.{Failure, Success, Try, Using}
+
+/** Responsible for the management of caches associated with a given data.
+  * @tparam T the type of the underlying cached data entry
+  */
+trait Cache[T] {
+
+  def logLevel: Level
+
+  import Cache._
+
+  /** Type of Metadata representation corresponding to the given data
+    */
+  type M <: Metadata
+
+  /** Saves the provided IR into the cache for this module.
+    *
+    * @param entry the data representation to be saved
+    * @param context the language context in which saving is taking place
+    * @param useGlobalCacheLocations if true, will use global caches' location, local one otherwise
+    * @return returns the location of the cache if successful, and [[None]] if
+    *         it was unable to save
+    */
+  def save(
+    entry: T,
+    context: EnsoContext,
+    useGlobalCacheLocations: Boolean
+  ): Option[TruffleFile] = {
+    implicit val logger: TruffleLogger = context.getLogger(this.getClass)
+    getCacheRoots(context) match {
+      case Some(roots) =>
+        if (useGlobalCacheLocations) {
+          if (saveCacheTo(roots.globalCacheRoot, entry)) {
+            return Some(roots.globalCacheRoot)
+          }
+        } else {
+          logger.log(
+            logLevel,
+            s"Skipping use of global cache locations for $stringRepr."
+          )
+        }
+
+        if (saveCacheTo(roots.localCacheRoot, entry)) {
+          return Some(roots.localCacheRoot)
+        }
+
+        logger.log(
+          logLevel,
+          s"Unable to write cache data for $stringRepr."
+        )
+        None
+      case None => None
+    }
+  }
+
+  protected def computeDigest(entry: T): Option[String]
+
+  protected def computeDigestFromSource(context: EnsoContext): Option[String]
+
+  protected def getCacheRoots(context: EnsoContext): Option[Roots]
+
+  protected def extractObjectToSerialize(entry: T): Object
+
+  /** Saves the cache into the provided `cacheRoot`.
+    *
+    * @param cacheRoot the root into which the cache should be saved
+    * @param entry    the data that should be cached
+    * @param logger   a logger
+    * @return `true` if the cache was written successfully, `false` otherwise
+    */
+  private def saveCacheTo(
+    cacheRoot: TruffleFile,
+    entry: T
+  )(implicit logger: TruffleLogger): Boolean = {
```

Comment:
Out of curiosity? Is there any reason/benefit of using `implicit logger` over making it normal method argument?

Reply:
runtime - no. It's only that you don't have to pass it around in the code because it is inserted implicitly by the compiler.